java/ql/automodel/publish.sh

```diff
@@ -1,31 +1,118 @@
-#!/bin/sh
+#!/bin/bash
 set -e
 
-# Before running this, make sure there is an SSO-enabled token with package:write
-# permissions to codeql supplied via the GITHUB_TOKEN environment variable
+# Add help message
+if [ "$1" = "-h" ] || [ "$1" = "--help" ]; then
+  echo "Usage: ./publish [override-release]"
+  echo "By default we publish the version of the codeql repo specified by the latest official release defined by the codeml-automodel repo."
+  echo "Otherwise, the optional argument override-release forces your current HEAD to be published."
+  exit 0
+fi
+
+# If we're publishing the codeml-automodel release then we will checkout the sha specified in the release.
+# So we need to check that there are no uncommitted changes in the local branch.
+# And, if we're publishing the current HEAD, it's cleaner to ensure that there are no uncommitted changes.
+if ! git diff --quiet; then
+  echo "Error: Uncommitted changes exist. Please commit or stash your changes before publishing."
+  exit 1
+fi
+
+# Check the above environment variables are set
+if [ -z "${GITHUB_TOKEN:-}" ]; then
+  echo "Error: GITHUB_TOKEN environment variable not set. Please set this to a token with package:write permissions to codeql."
+  exit 1
+fi
+if [ -z "${CODEQL_DIST:-}" ]; then
+  echo "Error: CODEQL_DIST environment variable not set. Please set this to the path of a codeql distribution."
+  exit 1
+fi
+if [ -z "${GH_TOKEN:-}" ]; then
+  echo "Error: GH_TOKEN environment variable not set. Please set this to a token with repo permissions to github/codeml-automodel."
+  exit 1
+fi
+
+# Get the sha of the previous release
+PREVIOUS_RELEASE_SHA=$(git rev-list -n 1 main -- ./src/qlpack.yml)
+if [ -z "$PREVIOUS_RELEASE_SHA" ]; then
+  echo "Error: Could not get the sha of the previous release of codeml-automodel query pack"
+  exit 1
+else
+  echo "Previous query-pack release sha: $PREVIOUS_RELEASE_SHA"
+fi
+
+CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+CURRENT_SHA=$(git rev-parse HEAD)
+
+if [ -z "${1:-}" ]; then
```

Comment:
I think it's simpler to type:
`./publish`
only. Because only in exceptional cases do we want to override. And so, for those exceptional cases, you have to specify an extra parameter.

Reply:
Yes I agree - I was just making a comment on how to parse the presence/absence of that parameter

Reply:
Fixed.

Reply:
Not quite. There's still no check that at most 1 argument is supplied. My suggestion is the way to do it.

Reply:
Woops, sorry, rushing.

Reply:
OK, fixed now (and I tested it).